src/crush/CrushWrapper.cc

```diff
@@ -1230,6 +1230,46 @@ int CrushWrapper::remove_rule(int ruleno)
   return 0;
 }
 
+int CrushWrapper::update_device_class(CephContext *cct, int id, const string& class_name, const string& name)
+{
+  int class_id = get_class_id(class_name);
+  if (class_id < 0) {
+    ldout(cct, 0) << "update_device_class class " << class_name << " does not exist " << dendl;
+    return -ENOENT;
+  }
+  if (id < 0) {
+    ldout(cct, 0) << "update_device_class " << name << " id " << id << " is negative " << dendl;
+    return -EINVAL;
+  }
+  assert(item_exists(id));
+
+  if (class_map.count(id) != 0 && class_map[id] == class_id) {
+    ldout(cct, 5) << "update_device_class " << name << " already set to class " << class_name << dendl;
+    return 0;
+  }
+
+  if (class_map.count(id)) {
+    int old_class_id = class_map[id];
+    int class_id_count = 0;
+    for (auto &c : class_map)
+      if (c.second == old_class_id)
+	class_id_count++;
+    if (class_id_count < 2) {
+      ldout(cct, 0) << "update_device_class " << name << " id " << id << " is the last device "
+		    << "with class " << class_name[old_class_id] << ", manually edit the crushmap "
+		    << "to remove the class entirely" << dendl;
```

Comment:
> This is the only bit that worries me. If i set a class wrong (e.g. typo) i can't fix it.

set-device-class will not create a new class, it will only set the class of a device if it already exists

> Is the concern that you'll end up with an empty device tree? (This seems fine to me.) Or no device tree at all?

If this command implies creation / deletion of classes, it will have to handle all the associated border cases. If it can only be used to set the class of a device when the class already exists and when doing so does not obsolete the class entirely, it is much simpler. For instance, removing the last device of a class means the associated class bucket tree is removed permanently and the rules that use it point to a non existent tree and the pool using this rule no longer reference something that exists.

From a sysadmin point of view it does not seem unreasonable to require that some operations such as creating classes, removing classes are done by editing the crushmap. And only operations that are requirement for daily cluster management are implemented via the cli (such as setting or changing the class of a given device).

I did not think this through though and maybe the cli should allow the full range of operations so that the admin is never required to edit the crushmap. 

What do you think ?